File: LC_23.py
# 23. Merge k Sorted Lists

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# TC: O(n log k)
# SC: O(n)
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        ListNode.__lt__ = lambda x, y: True if x.val < y.val else False
        
        dummy = ListNode()
        cur = dummy
        heap = []
        
        for head in lists:
            if head:
                heap.append(head)
        
        heapq.heapify(heap)
        
        while heap:
            node = heapq.heappop(heap)
            cur.next = node
            if node.next:
                heapq.heappush(heap, node.next)
            cur = cur.next
            
        return dummy.next

File: LC_23.java
// 23. Merge k Sorted Lists

// TC: O(n log k)
// SC: O(n)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> priorityQueue = 
            new PriorityQueue<>(new Comparator<ListNode>() {
                public int compare(ListNode l1, ListNode l2) {
                    return l1.val - l2.val;
                }
            });
        ListNode dummy = new ListNode();
        ListNode cur = dummy;
        for (ListNode head : lists) {
            if (head != null) {
                priorityQueue.add(head);
            }
        }
        
        while (!priorityQueue.isEmpty()) {
            ListNode node = priorityQueue.poll();
            cur.next = node;
            cur = cur.next;
            if (node.next != null) {
                priorityQueue.add(node.next);
            }
        }
        
        return dummy.next;
    }    
}
